src/libraries/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs

```diff
@@ -9,88 +10,188 @@ internal struct FileStatus
 
-            return ((_fileStatus.Mode & (int)readBit) != 0 && // has read permission
-                (_fileStatus.Mode & (int)writeBit) == 0);     // but not write permission
+        internal static void Initialize(ref FileStatus status, bool initiallyDirectory)
+        {
+            status.InitiallyDirectory = initiallyDirectory;
+            status.InvalidateCaches();
         }
 
-        public FileAttributes GetAttributes(ReadOnlySpan<char> path, ReadOnlySpan<char> fileName)
+        // Sets the cache initialization flags to -1, which means the caches are now uninitialized
+        internal void InvalidateCaches()
+        {
+            _initializedFileCache = -1;
+            _initializedSymlinkCache = -1;
+        }
+
+        internal bool IsReadOnly(ReadOnlySpan<char> path, bool continueOnError = false)
+        {
+            EnsureCachesInitialized(path, continueOnError);
+            return HasReadOnlyFlag;
+        }
+
+        internal bool IsHidden(ReadOnlySpan<char> path, bool continueOnError = false)
         {
-            // IMPORTANT: Attribute logic must match the logic in FileSystemEntry
+            EnsureCachesInitialized(path, continueOnError);
+            return HasHiddenFlag;
+        }
+
+        // Returns true if the path points to a directory, or if the path is a symbolic link
+        // that points to a directory
+        internal bool IsDirectory(ReadOnlySpan<char> path, bool continueOnError = false)
+        {
+            EnsureCachesInitialized(path, continueOnError);
+            return _isDirectory;
+        }
 
-            EnsureStatInitialized(path);
+        internal bool IsSymbolicLink(ReadOnlySpan<char> path, bool continueOnError = false)
+        {
+            EnsureCachesInitialized(path, continueOnError);
+            return HasSymbolicLinkFlag;
+        }
+
+        public FileAttributes GetAttributes(ReadOnlySpan<char> path, ReadOnlySpan<char> fileName)
+        {
+            EnsureCachesInitialized(path);
 
             if (!_exists)
                 return (FileAttributes)(-1);
 
             FileAttributes attributes = default;
 
-            if (IsReadOnly(path))
+            if (HasReadOnlyFlag)
                 attributes |= FileAttributes.ReadOnly;
 
-            if ((_fileStatus.Mode & Interop.Sys.FileTypes.S_IFMT) == Interop.Sys.FileTypes.S_IFLNK)
+            if (HasSymbolicLinkFlag)
                 attributes |= FileAttributes.ReparsePoint;
 
-            if (_isDirectory)
+            if (_isDirectory) // Refresh caches this
                 attributes |= FileAttributes.Directory;
 
-            // If the filename starts with a period or has UF_HIDDEN flag set, it's hidden.
-            if (fileName.Length > 0 && (fileName[0] == '.' || (_fileStatus.UserFlags & (uint)Interop.Sys.UserFlags.UF_HIDDEN) == (uint)Interop.Sys.UserFlags.UF_HIDDEN))
```

Comment:
@iSazonov, you wrote in the other comment:

> _I'm afraid it's completely incorrect to map extended attributes to classic attributes._
> _I think we need new ExtendedAttributes (or something like) on both Unix and Windows._

I completely agree with you that we should have an `ExtendedAttributes` property. It would solve the problem of having them intermixed with normal attributes.

Extended attributes apply to [BSD](https://www.unix.com/man-page/mojave/2/getattrlist/), [Linux](https://man7.org/linux/man-pages/man2/listxattr.2.html), [OSX](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/setxattr.2.html) and even for Windows, in the form of [Reparse Points](https://docs.microsoft.com/en-us/windows/win32/fileio/reparse-points).

But as you can see, `UF_HIDDEN` was already being read in the old code before my PR, and I just made sure it was consistently retrieved in `FileSystemInfo`, `FileSystemEntry` and `FileStatus`.

Reply:
I agree that it's not an either-or scenario: A platform-specific `ExtendedAttributes` property definitely makes sense, but it also makes sense to map the `UF_HIDDEN` flag to a platform-abstracted `Hidden` attribute (for instance, to the macOS Finder (file manager) both `UF_HIDDEN` files and files that observe the general Unix name-based convention (name starts with `.`) are hidden.

Has this `ExtendedAttributes` property been discussed at all? 

Note that macOS has _open-ended_ extended file-system attributes, similar to NTFS alternate streams (use `xattr -s -l /tmp` to see an example on macOS), but - if I understand correctly - NTFS alternate streams aren't supported (yet) in .NET, so that's probably a discussion for another day.

Reply:
> _Has this ExtendedAttributes property been discussed at all?_

Yes, we have a relatively new issue tracking this request: https://github.com/dotnet/runtime/issues/49604

Let's take the conversation there.

Reply:
>But as you can see, UF_HIDDEN was already being read in the old code before my PR, and I just made sure it was consistently retrieved in FileSystemInfo, FileSystemEntry and FileStatus.

It was added at 5.0 milestone without in-depth analysis. 
I'd suggest to pass the design through in-depth analysis and API review.
My main concern is that it (chflags) is **absolutely another API** with specific behaviors and specific error codes.
I have not found a _standard_ for these attributes. The most comprehensive list is https://www.freebsd.org/cgi/man.cgi?query=chflags&sektion=2&manpath=freebsd-release-ports
As we can see there are references to DOS, Windows and CIFS FILE_ATTRIBUTE_* attributes. I don't understand what that means and what the design intentions is.
_If there are no standard behavior definitions, I would make different APIs to avoid intractable problems in the future._

Reply:
After taking a look at #49604 it occurred to me that what we are talking about here is _not_ a case of  _extended attributes_, as the latter are user-definable data with _no special meaning to the file-system or the system_.

By contrast, the so-called _file flags_ on macOS / FreeBSD - of which `UF_HIDDEN` is one example - _do_ have system-defined meaning, analogous to the _file attributes_ on Windows.

These file flags do not map cleanly onto the born-on-Windows [`System.IO.FileAttributes`](https://docs.microsoft.com/en-US/dotnet/api/System.IO.FileAttributes) enumeration, though mapping `UF_HIDDEN`  onto `Hidden` is currently _partially_ implemented:

Examining an enumerated `FileSystemInfo` instance does surface `UF_HIDDEN` as the `Hidden` attribute in `.Attributes`, but the enumeration itself does _not_ consider such files hidden.

Curiously, the same inconsistency applies to the `ls` utility on macOS (my _guess_ is that the following also applies to FreeBSD): `ls` lists `UF_HIDDEN` files as if they _weren't_ hidden, but `ls -lO` does list all file flags (including `UF_HIDDEN` as `hidden`, the name also used by the `chflags` utility).
When POSIX-compatible shells on macOS perform globbing, they also do not consider `UF_HIDDEN` files hidden.

By contrast, Finder, the macOS file manager, equally considers `.`-initial names and `UF_HIDDEN` files hidden.

De facto the `.Attributes` property _is_ already being used to map non-Windows concepts onto the _equivalent_ Windows concepts:

* Symlinks on Unix-like platforms get the pseudo attribute `ReparsePoint`
* The user-specific lack of write permissions is mapped on to pseudo attribute `ReadOnly`

To me, these mappings make sense, but I agree that they should be thoroughly reviewed, with a view toward _extending_ this mapping; e.g., `UF_IMMUTABLE` and `SF_IMMUTABLE` are candidates for mapping onto `ReadOnly` too.

With respect to `UF_HIDDEN`, we need to decided whether (a) we want to resolve the current inconsistency and, if so, (b) in what direction.  

Consistently mapping both `.`-initial names and `UF_HIDDEN` files onto `Hidden` is defensible, even though in the world of shells it would make PowerShell's behavior then differ from that of `ls` / POSIX-compatible shells.

>  then surely this API must perform "." mapping in UF_HIDDEN

I don't think so: it seems fine to me to map into one direction: to have the platform-abstracted `Hidden` attribute reflect whether the file is effectively hidden, _irrespective of the mechanism used_, and on macOS / FreeBSD there happen to be two mechanisms available, both of which are - unfortunately only _partly_ - honored by the system.

If we were to introduce a dedicated API for macOS / FreeBSD _file flags_ (again, this is _not_ the same as _extended attributes_), then it should obviously reflect the true flag values only.

Reply:
> After taking a look at #49604 it occurred to me that what we are talking about here is _not_ a case of _extended attributes_, as the latter are user-definable data with _no special meaning to the file-system or the system_.

Yes, I suggest to use another terminology here - file system flags/ fsflags/ FileSystemInfo.Flags.

Since mapping is very questionable and may cause breaking conflicts in the future, I suggest avoiding it now and only presenting it in some way.
The application can be fixed faster than .Net. For example, PowerShell will be able to handle UF_HIDDEN in a special way and this can be changed in any intermediate version, unlike .Net which will remain stable from version to version.